docs/howtos/optax_update_guide.rst

```diff
@@ -0,0 +1,269 @@
+Upgrading my Codebase to Optax
+==============================
+
+We have proposed to replace :py:mod:`flax.optim` with `Optax
+<https://optax.readthedocs.io>`_ in 2021 with `FLIP #1009
+<https://github.com/google/flax/blob/main/docs/flip/1009-optimizer-api.md>`_ and
+the Flax optimizers are now *effectively deprecated*. This guide is targeted
+towards :py:mod:`flax.optim` users to help them update their code to Optax.
+
+Code samples below are executable in
+`Colab <https://colab.research.google.com/github/google/flax/blob/main/docs/notebooks/optax_update_guide.ipynb>`_.
+
+See also Optax's quick start documentation:
+https://optax.readthedocs.io/en/latest/optax-101.html
+
+.. testsetup::
+
+  import flax
+  import jax
+  import jax.numpy as jnp
+  import flax.linen as nn
+  import optax
+
+  # Note: this is the minimal code required to make below code run. See in the
+  # Colab linked above for a more meaningful definition of datasets etc. 
+  batch = {'image': jnp.ones([1, 28, 28, 1]), 'label': jnp.array([0])}
+  ds_train = [batch]
+  get_ds_train = lambda: [batch]
+  model = nn.Dense(1)
+  variables = model.init(jax.random.PRNGKey(0), batch['image'])
+  learning_rate, momentum, weight_decay, grad_clip_norm = .1, .9, 1e-3, 1.
+  loss = lambda params, batch: jnp.array(0.)
+
+Replacing ``flax.optim`` with ``optax``
+---------------------------------------
+
+Optax has drop-in replacements for all of Flax's optimizers. Refer to Optax's
+documentation `Common Optimizers <https://optax.readthedocs.io/en/latest/api.html>`_
+for API details.
+
+The usage is very similar, with the difference that ``optax`` does not keep a
+copy of the ``params``, so they need to be passed around separately. Flax
+provides the utility :py:class:`~flax.training.train_state.TrainState` to store
+optimizer state, parameters, and other associated data in a single dataclass.
+
+.. codediff::
```

Comment:
Very cool you used the side-by-side view, great to see that it is actually useful 😄 . Not sure if you are aware of this, but if you add a "`#!`" to the end of a line, that line will be highlighted. (I can also see that it may not be necessary in this HOWTO, so feel free to ignore.)

Reply:
Ah, I was wondering why no lines were highlighted (I somehow assumed it would be automatic).

Good to know, but I went through the diffs (or rather "side by sides") and didn't really find value in aligning the code snippets and highlighting the changes (there are too few common lines). But good to know for my next HOTWO.